packages/codemirror-blocks/src/ui/PrimitiveList.tsx

```diff
@@ -93,69 +93,139 @@ export const PrimitiveGroup = (props: PrimitiveGroupProps) => {
-      group={g}
-      onFocus={onFocus}
-      onKeyDown={onKeyDown}
-      selected={selected}
-    />
-  );
-  const renderPrimitive = (p: LanguagePrimitive) => (
-    <Primitive
-      key={p.name}
-      primitive={p}
-      onFocus={onFocus}
-      onKeyDown={onKeyDown}
-      className={selected == p.name ? "selected" : ""}
-    />
-  );
+export const PrimitiveList = React.forwardRef(
+  (props: PrimitiveListProps, ref: React.MutableRefObject<HTMLElement>) => {
+    const {
+      primitives = [],
+      selected,
+      setSelectedPrimitive,
+      selectedPrimitive,
+      toolbarRef,
+      searchString,
+    } = props;
 
-  const text = searchString
-    ? (primitives.length == 0 ? "No" : primitives.length) + " blocks found"
-    : "blocks";
+    // @pcardune - typing this array has been a challenge
+    // JSX.Element doesn't work, even though that's the
+    // recommendation I'm seeing online
+    const primitiveRefs = useRef<any[]>([]);
 
-  return (
-    <div>
-      <h3
-        id="toolbar_heading"
-        className="screenreader-only"
-        aria-live="assertive"
-        aria-atomic="true"
-      >
-        {text}
-      </h3>
-      <ul
-        className="PrimitiveList list-group"
-        aria-labelledby="toolbar_heading"
-      >
-        {primitives.map((item) =>
-          item instanceof PrimitiveGroupModel
-            ? renderGroup(item)
-            : renderPrimitive(item)
-        )}
-      </ul>
-    </div>
-  );
-};
+    const renderGroup = (g: PrimitiveGroupModel) => (
+      <PrimitiveGroup
+        key={g.name}
+        group={g}
+        selected={selected}
+        selectedPrimitive={selectedPrimitive}
+        toolbarRef={toolbarRef}
+        setSelectedPrimitive={setSelectedPrimitive}
+      />
+    );
+
+    const renderPrimitive = (p: LanguagePrimitive, i: number) => {
+      const primRef = React.createRef<HTMLElement>();
+      const prim = (
+        <Primitive
+          key={p.name}
+          primitive={p}
+          onFocus={() => setSelectedPrimitive(p)}
+          className={selected == p.name ? "selected" : ""}
+          ref={primRef}
+        />
+      );
+      primitiveRefs.current[i] = prim;
+      return prim;
+    };
+
+    // Set selectedPrimitive state, depending on whether we go up or down
+    const move = (event: React.KeyboardEvent, dir: "Up" | "Down") => {
+      if (!selectedPrimitive || primitives.length == 0) {
+        return; // Nothing to select. Bail.
+      }
+
+      // compute the index of the newly-selected primitive, and get the Element
+      let newIndex;
+      const prevIndex = primitives.indexOf(selectedPrimitive); // -1 if nothing selected
+      if (dir == "Down") {
+        newIndex = Math.min(prevIndex + 1, primitives.length - 1);
+      } else {
+        newIndex = Math.max(prevIndex - 1, 0);
+      }
+      const newPrimitiveElt = primitiveRefs.current[newIndex];
+
+      // @pcardune - is it safe to assume this will always exist, and ditch the check?
```

Comment:
The only edge case I can think of is if the list of primitives gets shorter and you index off the end of the array, but then the index you would be using would be different anyway so I think it's fine.